velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,223 @@
+  return sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+}
+
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize(const TypeKind& typeKind) {
+  return VELOX_DYNAMIC_TYPE_DISPATCH(normalizeSize, typeKind);
+}
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn<TypeKind::BIGINT>(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  using T = TypeTraits<TypeKind::BIGINT>::NativeType;
+  // Store null as min/max value according compare flags.
+  if (RowContainer::isNullAt(row, rowColumn.nullByte(), rowColumn.nullMask())) {
+    CompareFlags compareFlags = prefixSortLayout.compareFlag(index);
+    PrefixSortEncoder::encode(
+        ((compareFlags.ascending && compareFlags.nullsFirst) ||
+         (!compareFlags.ascending && !compareFlags.nullsFirst))
+            ? std::numeric_limits<T>::min()
+            : std::numeric_limits<T>::max(),
+        prefix + prefixSortLayout.prefixOffset(index));
+  } else {
+    PrefixSortEncoder::encode(
+        *(reinterpret_cast<T*>(row + rowColumn.offset())),
+        prefix + prefixSortLayout.prefixOffset(index));
+  }
+  // Invert bits if desc
+  if (!prefixSortLayout.compareFlag(index).ascending) {
+    for (uint64_t s = 0; s < sizeof(T); s++) {
+      *(prefix + prefixSortLayout.prefixOffset(index) + s) =
+          ~*(prefix + prefixSortLayout.prefixOffset(index) + s);
+    }
+  }
+}
+} // namespace
+
+PrefixSortLayout::PrefixSortLayout(
+    const std::vector<TypePtr>& types,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const uint32_t maxPrefixKeyLength)
+    : normalizedKeySize_(0),
+      numNormalizedKeys_(0),
+      numSortKeys_(types.size()),
+      keyCompareFlags_(keyCompareFlags) {
+  VELOX_CHECK_GE(types.size(), 0);
+  for (TypePtr type : types) {
+    if (numNormalizedKeys_ > maxPrefixKeyLength) {
+      break;
+    }
+    if (type->kind() == TypeKind::BIGINT) {
+      numNormalizedKeys_++;
+      normalizedKeySize_ += sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+    } else {
+      break;
+    }
+  }
+  entrySize_ = normalizedKeySize_ + sizeof(char*);
+  hasNonNormalizedKey_ = numNormalizedKeys_ < numSortKeys_;
+  // Compute prefix offsets for sort columns.
+  uint32_t offset = 0;
+  for (int i = 0; i < numNormalizedKeys_; i++) {
+    prefixOffsets_.push_back(offset);
+    offset += normalizeSize(types[i]->kind());
+  }
+}
+
+int PrefixSort::compareAllNormalizedKeys(char* left, char* right) {
+  // TODO : Support compare by words as a separate PR，benchmark performance
```

Comment:
Why is this a TODO? What prevents from addressing it now?

Reply:
emm...
Only a little problem with write padding logic ut for short/byte types  as we do not support them now,  and as you said:
>This is nice. Maybe open a separate PR for that change.

I think I can add it in this PR.

Reply:
done